Unxpected StringIndexOutOfBoundsException

**Describe the bug**
After migration to the 2022.0.3 version following exceptions have been appearing:
```
Caused by: java.lang.StringIndexOutOfBoundsException: begin 183, end 153, length 198
        at java.base/java.lang.String.checkBoundsBeginEnd(String.java:4602) ~[na:na]
        at java.base/java.lang.String.substring(String.java:2705) ~[na:na]
        at org.springframework.cloud.config.client.ConfigServerConfigDataLoader.extractApplicationName(ConfigServerConfigDataLoader.java:223) ~[spring-cloud-config-client-4.0.3.jar:4.0.3]
        at org.springframework.cloud.config.client.ConfigServerConfigDataLoader.isProfileSpecificPropertySource(ConfigServerConfigDataLoader.java:207) ~[spring-cloud-config-client-4.0.3.jar:4.0.3]
        at org.springframework.cloud.config.client.ConfigServerConfigDataLoader.lambda$doLoad$1(ConfigServerConfigDataLoader.java:160) ~[spring-cloud-config-client-4.0.3.jar:4.0.3]
```
Here are some clues:
- application.yaml
```
spring:
  config:
    import: configserver:http://dbs02inplayamq9.donbest.com:8888/
```
- debug values: ConfigServerConfigDataLoader#extractApplicationName
`propertySource.getName()` 
```
"configserver:ssh://git@stash.int.openbet.com:7999/dbs/environments.git/Config resource 'file [/tmp/config-repo-16512912790018624282/platform/pinnacle-sink.yaml' via location 'platform/' (document #0)"
```
- something wrong here:
```
		return !applicationName.equals(extractApplicationName(propertySource))
				&& propertySource.getName().matches(".*[-,]" + profile + ".*");
```

--- a/spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/ConfigServerConfigDataLoader.java
+++ b/spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/ConfigServerConfigDataLoader.java
@@ -24,7 +24,6 @@ import java.util.HashMap;
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
-import java.util.stream.Collectors;
 
 import org.apache.commons.logging.Log;
 
@@ -49,7 +48,6 @@ import org.springframework.http.HttpMethod;
 import org.springframework.http.HttpStatus;
 import org.springframework.http.MediaType;
 import org.springframework.http.ResponseEntity;
-import org.springframework.util.ObjectUtils;
 import org.springframework.util.StringUtils;
 import org.springframework.web.client.HttpClientErrorException;
 import org.springframework.web.client.HttpServerErrorException;
@@ -153,15 +151,28 @@ public class ConfigServerConfigDataLoader implements ConfigDataLoader<ConfigServ
 					}
 					else if (ALL_OPTIONS.size() > 2) {
 						// boot 2.4.5+
-						if (resource.isProfileSpecific()) {
-							List<PropertySource<?>> filteredSources = propertySources.stream()
-									.filter(propertySource -> resource.getAcceptedProfiles().stream()
-											.anyMatch(profile -> isProfileSpecificPropertySource(
-													resource.getProperties().getName(), propertySource, profile)))
-									.collect(Collectors.toList());
-							return new ConfigData(filteredSources, Option.IGNORE_IMPORTS, Option.PROFILE_SPECIFIC);
-						}
-						return new ConfigData(propertySources, Option.IGNORE_IMPORTS);
+						return new ConfigData(propertySources, propertySource -> {
+							String propertySourceName = propertySource.getName();
+							List<Option> options = new ArrayList<>();
+							options.add(Option.IGNORE_IMPORTS);
+							options.add(Option.IGNORE_PROFILES);
+							// TODO: the profile is now available on the backend
+							// in a future minor, add the profile associated with a
+							// PropertySource see
+							// https://github.com/spring-cloud/spring-cloud-config/issues/1874
+							for (String profile : resource.getAcceptedProfiles()) {
+								// TODO: switch to match
+								// , is used as a profile-separator for property sources
+								// from vault
+								// - is the default profile-separator for property sources
+								// TODO This is error prone logic see https://github.com/spring-cloud/spring-cloud-config/issues/2291
+								if (propertySourceName.matches(".*[-,]" + profile + ".*")) {
+									// // TODO: switch to Options.with() when implemented
+									options.add(Option.PROFILE_SPECIFIC);
+								}
+							}
+							return ConfigData.Options.of(options.toArray(new Option[0]));
+						});
 					}
 				}
 			}
@@ -190,36 +201,6 @@ public class ConfigServerConfigDataLoader implements ConfigDataLoader<ConfigServ
 		logger.warn("Could not locate PropertySource (" + resource + "): "
 				+ (error != null ? error.getMessage() : errorBody));
 		return null;
-	}
-
-	private boolean isProfileSpecificPropertySource(String applicationName, PropertySource<?> propertySource,
-			String profile) {
-		// Application names can have - so before checking if the application name
-		// contains a -
-		// check to see if the application name matches the property source name, if so
-		// its not a profile specific property source.
-
-		// TODO: switch to match
-		// , is used as a profile-separator for property sources
-		// from vault
-		// - is the default profile-separator for property sources
-		return !applicationName.equals(extractApplicationName(propertySource))
-				&& propertySource.getName().matches(".*[-,]" + profile + ".*");
-	}
-
-	private String extractApplicationName(PropertySource<?> propertySource) {
-		if (ObjectUtils.isEmpty(propertySource.getName())) {
-			return "";
-		}
-		int lastSlash = propertySource.getName().lastIndexOf("/");
-		int lastPeriod = propertySource.getName().lastIndexOf(".");
-		if (lastSlash == -1) {
-			lastSlash = 0;
-		}
-		if (lastPeriod == -1) {
-			lastPeriod = propertySource.getName().length();
-		}
-		return propertySource.getName().substring(lastSlash + 1, lastPeriod);
 	}
 
 	protected void log(Environment result) {

--- a/spring-cloud-config-client/src/test/java/org/springframework/cloud/config/client/ConfigServerConfigDataCustomizationIntegrationTests.java
+++ b/spring-cloud-config-client/src/test/java/org/springframework/cloud/config/client/ConfigServerConfigDataCustomizationIntegrationTests.java
@@ -124,7 +124,7 @@ public class ConfigServerConfigDataCustomizationIntegrationTests {
 				assertThat(options.contains(Option.PROFILE_SPECIFIC)).isFalse();
 			}
 			else {
-				assertThat(configData.getPropertySources()).hasSize(0);
+				assertThat(configData.getPropertySources()).hasSize(1);
 			}
 			return configData;
 		}

--- a/spring-cloud-config-client/src/test/java/org/springframework/cloud/config/client/ConfigServerConfigDataLoaderTests.java
+++ b/spring-cloud-config-client/src/test/java/org/springframework/cloud/config/client/ConfigServerConfigDataLoaderTests.java
@@ -20,8 +20,10 @@ import java.util.Arrays;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 
 import org.apache.commons.logging.Log;
+import org.junit.jupiter.api.Disabled;
 import org.junit.jupiter.api.Test;
 import org.mockito.ArgumentMatchers;
 
@@ -65,7 +67,9 @@ class ConfigServerConfigDataLoaderTests {
 
 	}
 
+	@Disabled
 	@Test
+	// TODO Enable once we have https://github.com/spring-cloud/spring-cloud-config/issues/2291
 	void filterPropertySourcesThatAreNotProfileSpecific() {
 		PropertySource p1 = new PropertySource("p1", Collections.singletonMap("foo", "bar"));
 		PropertySource p2 = new PropertySource("p2", Collections.singletonMap("hello", "world"));
@@ -74,7 +78,9 @@ class ConfigServerConfigDataLoaderTests {
 
 	}
 
+	@Disabled
 	@Test
+	// TODO Enable once we have https://github.com/spring-cloud/spring-cloud-config/issues/2291
 	void returnPropertySourcesThatAreProfileSpecific() {
 		PropertySource p1 = new PropertySource("p1-dev", Collections.singletonMap("foo", "bar"));
 		PropertySource p2 = new PropertySource("p2-dev", Collections.singletonMap("hello", "world"));
@@ -85,6 +91,48 @@ class ConfigServerConfigDataLoaderTests {
 				.contains(ConfigData.Option.IGNORE_IMPORTS)).isTrue();
 		assertThat(configData.getOptions(configData.getPropertySources().get(1))
 				.contains(ConfigData.Option.IGNORE_IMPORTS)).isTrue();
+
+	}
+
+	@Disabled
+	@Test
+	// TODO Enable once we have https://github.com/spring-cloud/spring-cloud-config/issues/2291
+	void filterPropertySourcesWithDocuments() {
+		PropertySource p1 = new PropertySource(
+			"configserver:git@github.com:demo/support-configuration-repo.git/application.yml",
+			Collections.singletonMap("foo", "bar"));
+		PropertySource p2 = new PropertySource(
+			"configserver:git@github.com:demo/support-configuration-repo.git/application-foo.yml",
+			Collections.singletonMap("foo", "bar"));
+		PropertySource p3 = new PropertySource(
+			"configserver:git@github.com:demo/support-configuration-repo.git/Config resource 'file [/var/folders/k3/zv8hzdm17vv69j485fv3cf9r0000gp/T/config-repo-14772121892716396795/commons/application.properties' via location 'commons/' (document #0)",
+			Collections.singletonMap("hello", "world"));
+		PropertySource p4 = new PropertySource(
+			"configserver:git@github.com:demo/support-configuration-repo.git/Config resource 'file [/var/folders/k3/zv8hzdm17vv69j485fv3cf9r0000gp/T/config-repo-14772121892716396795/commons/application-foo.properties' via location 'commons/' (document #0)",
+			Collections.singletonMap("hello", "world"));
+		Map<String, Object> activatesOnProfileCamelCase = new HashMap<>();
+		activatesOnProfileCamelCase.put("spring.config.activate.onProfile", "foo");
+		PropertySource p5 = new PropertySource(
+			"configserver:git@github.com:demo/support-configuration-repo.git/Config resource 'file [/var/folders/k3/zv8hzdm17vv69j485fv3cf9r0000gp/T/config-repo-14772121892716396795/commons/application.properties' via location 'commons/' (document #1)",
+			activatesOnProfileCamelCase);
+		Map<String, Object> activatesOnProfile = new HashMap<>();
+		activatesOnProfile.put("spring.config.activate.on-profile", "foo");
+		PropertySource p6 = new PropertySource(
+			"configserver:git@github.com:demo/support-configuration-repo.git/Config resource 'file [/var/folders/k3/zv8hzdm17vv69j485fv3cf9r0000gp/T/config-repo-14772121892716396795/commons/application.properties' via location 'commons/' (document #2)",
+			activatesOnProfile);
+		PropertySource p7 = new PropertySource(
+			"configserver:git@github.com:demo/support-configuration-repo.git/application-foo.yaml",
+			Collections.singletonMap("hello", "world"));
+		PropertySource p8 = new PropertySource(
+			"configserver:git@github.com:demo/support-configuration-repo.git/Config resource 'file [/var/folders/k3/zv8hzdm17vv69j485fv3cf9r0000gp/T/config-repo-14772121892716396795/commons/application-foo.yaml' via location 'commons/' (document #0)",
+			Collections.singletonMap("hello", "world"));
+		PropertySource p9 = new PropertySource(
+			"configserver:git@github.com:demo/support-configuration-repo.git/Config resource 'file [/var/folders/k3/zv8hzdm17vv69j485fv3cf9r0000gp/T/config-repo-14772121892716396795/commons/application-foo.yaml' via location 'commons/' (document #1)",
+			Collections.singletonMap("hello", "world"));
+
+		ConfigData configData = setupConfigServerConfigDataLoader(Arrays.asList(p1, p2, p3, p4, p5, p6, p7, p8, p9),
+			"application-slash", "foo");
+		assertThat(configData.getPropertySources().size()).isEqualTo(7);
 
 	}
 

